### PR TITLE
Reduce `[]ConfigKey` slice allocate numbers during xds caching

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

This is making use  of sync.Pool to relieve pressure on the garbage collector.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Only need to look at the latest commit https://github.com/istio/istio/pull/39731/commits/0a6f271b2cce392994f4a8dd5b6b22c254a7bd1b

File: pilot/pkg/model/xds_cache.go [hunk 1]
```diff
@@ -164,20 +188,67 @@ func newLru(evictCallback simplelru.EvictCallback) simplelru.LRUCache {
...
+}
+
+// clearEvicted is to clear configIndex and typesIndex based on evictedKeys.gi
+func (l *LruCache) clearEvicted() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// In order to save cpu, if the evicted keys is very small, we can just ignore
+	// TODO: make it configurable
+	if len(l.evictedKeys) < 100 {
```
Comment (adiprerepa) [inline pilot/pkg/model/xds_cache.go:225]: is the optimization worth the inaccuracy? I think things like delta will depend on whether or not entries in the cache are still valid. We could just check the cache AND evictedKeys though -- @howardjohn

File: pilot/pkg/xds/endpoint_builder.go [hunk 2]
```diff
@@ -177,8 +179,13 @@ func (b EndpointBuilder) Cacheable() bool {
 	return b.service != nil
 }
 
-func (b EndpointBuilder) DependentConfigs() []model.ConfigKey {
-	configs := []model.ConfigKey{}
+func (b EndpointBuilder) DependentConfigs(allocator *model.Allocator) []model.ConfigKey {
+	var configs []model.ConfigKey
+	if allocator != nil {
+		configs = allocator.Allocate(2)
```
Comment (adiprerepa) [inline pilot/pkg/xds/endpoint_builder.go:185]: why 2 here?

File: pilot/pkg/model/xds_cache.go [hunk 3]
```diff
@@ -164,20 +188,67 @@ func newLru(evictCallback simplelru.EvictCallback) simplelru.LRUCache {
...
-func (l *lruCache) evict(k interface{}, v interface{}) {
+// This is the callback passed to LRU, it will be called whenever a key is removed.
+func (l *LruCache) onEvict(k interface{}, _ interface{}) {
 	if features.EnableXDSCacheMetrics {
 		xdsCacheEvictions.Increment()
 	}
 
-	key := k.(string)
-	value := v.(cacheValue)
+	if l.trackEvicted {
```
Comment (adiprerepa) [inline pilot/pkg/model/xds_cache.go:197]: in what scenario do we not want to track evicted? do we just not evict if this is not set?

**Review by adiprerepa — COMMENTED** [review]
optimization should help a lot

**Comment by adiprerepa** [comment]
@hzxuzhonghu should https://github.com/istio/istio/pull/39688 be closed in favor of this?

**Comment by hzxuzhonghu** [comment]
I am ok unless some others not want to couple them.

But the commit is largely dependent on previous pr.

File: pilot/pkg/xds/endpoint_builder.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/endpoint_builder.go:185]: At most 2 is needed for EndpointBuilder

File: pilot/pkg/model/xds_cache.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:197]: Actively `Remove`

File: pilot/pkg/model/xds_cache.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:225]: Not sure i understand

Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:225]: This can be seen as a debouncer, hope you can better understand

**Review by howardjohn — COMMENTED** [review]
As mentioned before, I don't feel comfortable merging these types of performance changes, which may or may not bring improvement, without numbers actually showing differences.

Currently our benchmarks only show a regression from this change, so if the are not giving accurate coverage then please update them.

**Comment by hzxuzhonghu** [comment]
i have to say can we be fair?  addind dependentconfigs to cache value does not improve any perf, but on opposite mem costing.. i just reverted the bad stuff in #39688

**Pull request closed without merging**